server/src/main/java/org/candlepin/controller/CandlepinPoolManager.java

```diff
@@ -2282,17 +2285,26 @@ public void deletePools(Collection<Pool> pools, Collection<String> alreadyDelete
                     this.enforcer.postUnbind(entitlement.getConsumer(), this, entitlement);
                 }
 
-                // Recalculate status for affected consumers
-                log.info("Recomputing status for {} consumers", consumerStackedEnts.size());
-                int i = 0;
-                for (Consumer consumer : consumerStackedEnts.keySet()) {
-                    this.complianceRules.getStatus(consumer);
+                /* Save the consumer IDs and then detach all the keys in the consumerStackEnts map.
+                 * Otherwise during the status calculations, the facts proxy objects objects will be resolved
+                 * and the memory use will grow linearly with the number of consumers instead of remaining
+                 * constant as we calculate the status of each consumer.  See BZ 1584259 */
+                List<String> consumers = consumerStackedEnts.keySet().stream()
+                    .map(Consumer::getId)
+                    .collect(Collectors.toList());
+                this.consumerCurator.batchDetach(consumerStackedEnts.keySet());
 
-                    if (++i % 1000 == 0) {
-                        this.consumerCurator.flush();
+                log.info("Recomputing status for {} consumers", consumers.size());
+
+                // Recalculate status for affected consumers
+                for (List<String> subList : Lists.partition(consumers, 1000)) {
+                    for (String consumerId : subList) {
+                        Consumer consumer = consumerCurator.get(consumerId);
```

Comment:
We briefly talked about this on Friday, but since we've evicted the consumers from the cache, this is going to hit the DB for every consumer again. If we changed this loop around a bit such that we can bulk up the consumer fetching for blocks of consumers, we'll save a bunch of time on DB hits/latency (in exchange for some of that memory we've been abusing):

```
for (/* partition loop */) {
    for (Consumer consumer : this.consumerCurator.getConsumers(subList)) {
        this.complianceRules.getStatus(consumer);
        this.consumerCurator.detach(consumer);
    }
    this.consumerCurator.flush();
}
```

If this is still heavy on memory, we could tune the block size accordingly; but anything higher than one should make a measurable difference.

Reply:
`getConsumers` doesn't actually exist.  Is there maybe a method in AbstractHibernateCurator you were thinking of?

Reply:
Nope, it'd have to be written, but it's fairly trivial to write. Should be no more than five lines of code since we're using the consumer ID.